core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -33,15 +35,19 @@
  */
 public class CompositeReadableBuffer extends AbstractReadableBuffer {
 
+  private final Deque<ReadableBuffer> readableBuffers;
+  private final Deque<ReadableBuffer> rewindableBuffers;
   private int readableBytes;
-  private final Queue<ReadableBuffer> buffers;
+  private boolean marked;
 
   public CompositeReadableBuffer(int initialCapacity) {
-    buffers = new ArrayDeque<>(initialCapacity);
+    readableBuffers = new ArrayDeque<>(initialCapacity);
+    rewindableBuffers = new ArrayDeque<>(initialCapacity);
   }
 
   public CompositeReadableBuffer() {
-    buffers = new ArrayDeque<>();
+    readableBuffers = new ArrayDeque<>();
+    rewindableBuffers = new ArrayDeque<>();
```

Comment:
This could be initialized to a constant empty queue, and replaced with a new ArrayDeque in the one place that it is appended to (in `advanceBuffer()`). It would save couple more allocations in cases that mark/reset isn't used.

Reply:
Alright, changed to instantiate `rewindableBuffers` lazily.